clang/include/clang/CIR/Dialect/IR/CIRAttrs.td

```diff
@@ -1249,7 +1250,150 @@ def GlobalAnnotationValuesAttr : CIR_Attr<"GlobalAnnotationValues",
   let genVerifyDecl = 1;
 }
 
+
+//===----------------------------------------------------------------------===//
+// TBAAAttr
+//===----------------------------------------------------------------------===//
+
 def CIR_TBAAAttr : CIR_Attr<"TBAA", "tbaa", []> {
+  let summary = "CIR dialect TBAA base attribute";
+}
+
+//===----------------------------------------------------------------------===//
+// TBAATypeDescriptorAttr
+//===----------------------------------------------------------------------===//
+
+def CIR_TBAAMemberAttr : CIR_Attr<"TBAAMember", "tbaa_member", []> {
+  let summary = "Attribute representing a member of a TBAA structured type.";
+
+  let parameters = (ins "TBAAAttr":$typeDesc,
+                        "int64_t":$offset,
+                        "int64_t":$size);
+
+  let builders = [
+    AttrBuilderWithInferredContext<(ins "TBAAAttr":$typeDesc,
+                                        "int64_t":$offset,
+                                        "int64_t":$size), [{
+      return $_get(typeDesc.getContext(), typeDesc, offset, size);
+    }]>
+  ];
+  let assemblyFormat = "`<` params `>`";
+}
+
+def CIR_TBAAMemberAttrArray : ArrayRefParameter<"TBAAMemberAttr"> {
+  let summary = "Array of TBAAMemberAttr attributes.";
+
+  let printer = [{
+    $_printer << '{';
+    llvm::interleaveComma($_self, $_printer, [&](TBAAMemberAttr attr) {
+        $_printer.printStrippedAttrOrType(attr);
+    });
+    $_printer << '}';
+  }];
+
+  let parser = [{
+    [&]() -> llvm::FailureOr<llvm::SmallVector<TBAAMemberAttr>> {
+        using Result = llvm::SmallVector<TBAAMemberAttr>;
+        if ($_parser.parseLBrace())
+            return mlir::failure();
+        llvm::FailureOr<Result> result = mlir::FieldParser<Result>::parse($_parser);
+        if (failed(result))
+            return mlir::failure();
+        if ($_parser.parseRBrace())
+            return mlir::failure();
+        return result;
+    }()
+  }];
+}
+
+class CIR_TBAATypeDescriptorAttr<string name, string attrMnemonic>: CIR_Attr<name, attrMnemonic, [], "TBAAAttr"> {
+  let summary = "Base class for TBAA type descriptor attributes.";
+}
+
+def CIR_TBAAScalarTypeDescriptorAttr : CIR_TBAATypeDescriptorAttr<"TBAAScalarTypeDescriptor",
+                                                 "tbaa_scalar_type_desc"> {
```

Comment:
indentation feels odd here, same for other some other ones

Reply:
Are there any tools available to format `.td` files? For example, just as we can use `clang-format` to format `.cpp` files:

```
clang-format -i /path/to/cpp_file
```

Reply:
Maybe? Last I followed was https://discourse.llvm.org/t/formating-mlir-tablegen-code/60767, but I'm not sure there has been any development.